File: C++/ProductArrayExceptSelf/ProductArrayExceptSelf.cpp
/* 238. Product of Array Except Self
Given an array nums of n integers where n > 1,  return an array output such that output[i] is equal to the product of all the elements of nums except nums[i].

Example:

Input:  [1,2,3,4]
Output: [24,12,8,6]

Note: Please solve it without division and in O(n).
 */

#include <iostream>
#include <vector>


void print(std::vector<int> vec)
{
	for (auto const &v : vec)
	{		
		std::cout << v << ' ';		
		
	}
	std::cout << std::endl;
}

std::vector<int> productExceptSelf(std::vector<int> &nums)
{
	std::vector<int> result;

	for(int i=0; i<nums.size(); i++)
	{
		
	}

	return result;
}

int main()
{

	std::vector<int> test = { 1,2,3,4 };

	productExceptSelf(test);

	//print(test);

	return 0;
}

File: Python/Product_Array_Except_Self.py
""" 238. Product of Array Except Self
Given an array nums of n integers where n > 1,  return an array output such that output[i] is equal to
the product of all the elements of nums except nums[i].

Example:
Input:  [1,2,3,4]
Output: [24,12,8,6]
Note: Please solve it without division and in O(n).
"""
from typing import List


class ProductArrayExceptSelf:
    @staticmethod
    def product_array(nums: List[int]) -> List[int]:
        result = []
        if not nums:
            return result
        # product result initialized to 1s
        result = [1 for _ in range(len(nums))]

        prev = 1
        for i in range(len(nums)):
            result[i] = prev
            prev *= nums[i]

        prev = 1
        for i in range(len(nums)-1, -1, -1):
            result[i] *= prev
            prev *= nums[i]

        return result


arr = [1, 2, 3, 4]
prod_arr = ProductArrayExceptSelf()
print(prod_arr.product_array(arr))
